snowflake/snowflake.go

```diff
@@ -0,0 +1,256 @@
+type SequenceResolver func(current int64) (uint16, error)
+
+// NodeIDProvider the snowflake Node Generator provider.
+type NodeIDProvider func() (uint16, error)
+
+// setEpoch set the start time for snowflake algorithm.
+// It will panic when:
+//   * t IsZero
+//   * t > current millisecond
+//   * current millisecond - t > 2^39( * 10ms).
+// NOTE: This function is thread-unsafe, call before init
+func (g *Generator) setEpoch(epoch time.Time) (err error) {
+	const op errors.Op = "snowflakeGenerator.setEpoch"
+	epoch = epoch.UTC()
+
+	if epoch.IsZero() {
+		err = errors.New(op, "Epoch cannot be a zero value")
+		return
+	}
+
+	if epoch.After(time.Now()) {
+		err = errors.New(op, "Epoch cannot be in the future")
+		return
+	}
+
+	// Because s must after now, so the `df` not < 0.
+	if since(epoch) > maxTimestamp {
+		err = errors.New(op, "The maximum life cycle of the snowflake algorithm is 179 years(2^39-10ms)")
+		return
+	}
+	g.epochGuard.Do(func() {
+		g.Epoch = epoch
+	})
+	return
+}
+
+// setResolver set the sequence custom Resolver
+func (g *Generator) setResolver(sequenceResolver SequenceResolver) (err error) {
+	const op errors.Op = "snowflakeGenerator.setResolver"
+	if sequenceResolver == nil {
+		err = errors.New(op, "SequenceResolver cannot be nil")
+		return
+	}
+	g.resolverGuard.Do(func() {
+		g.SequenceResolver = sequenceResolver
+	})
+	return
+}
+
+// setNodeIDProvider set the sequence NodeID provider
+func (g *Generator) setNodeIDProvider(nodeIDProvider NodeIDProvider) (err error) {
+	const op errors.Op = "snowflakeGenerator.setNodeIDProvider"
+	if nodeIDProvider == nil {
+		err = errors.New(op, "NodeIDProvider cannot be nil")
+		return
+	}
+	g.nodeIDProviderGuard.Do(func() {
+		g.NodeIDProvider = nodeIDProvider
+		var nodeID uint16
+		nodeID, err = g.NodeIDProvider()
+		if err != nil {
+			err = errors.New(op, "error generating nodeID", err)
+			return
+		}
+		g.nodeID = nodeID
+	})
+	return
+}
+
+// Decompose returns a map of snowflake id parts.
+func (g *Generator) Decompose(sid uint64) (id ID) {
+	id.MSB = int8(sid >> 63)
+	timestamp := sid >> timeShift
+	id.Timestamp = time.Unix(0, (g.Epoch.UnixNano())+int64(timestamp*timeUnit))
+	id.Sequence = uint16(sid & maskSequence)
+	id.NodeID = uint16(sid & maskNodeID >> nodeShift)
+	return
+}
+
+// NextID generate a new unique snowflake ID
+func (g *Generator) NextID() (sid uint64, err error) {
+	var current int64
+	var seq uint16
+	current, err = g.currentTimestamp()
+
+	if err != nil {
+		return
+	}
+
+	seq, err = g.SequenceResolver(current)
+
+	if err != nil {
+		return
+	}
+
+	for seq >= maxSequence {
+		current, err = g.waitForNext10Millis(current)
+		if err != nil {
+			return
+		}
+
+		seq, err = g.SequenceResolver(current)
+		if err != nil {
+			return
+		}
+	}
+	sid = uint64(current)<<timeShift | uint64(g.nodeID)<<nodeShift | uint64(seq)
+	return
+}
+
+func (g *Generator) currentTimestamp() (current int64, err error) {
+	const op errors.Op = "snowflake_id.currentTimestamp"
```

Comment:
I notice there is the inconsistency of `op` prefix `snowflake_id` vs `snowflakeGenerator`. I think keeping the same will create less confusion.

Reply:
Aha, my bad, need to fix this